### ISSUE TITLE
Error opening/closing files in riverpod repo

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#4295]
While investigating https://github.com/Dart-Code/Dart-Code/issues/3625#issuecomment-1333506945 from @rrousselGit:

```
1669889723876:Ex:An error occurred while handling textDocument/didClose notification
Invalid argument:: is a regular instance:: Instance of '_HashMap<dynamic, dynamic>'
#0      _SendPort._sendInternal (dart::isolate-patch/isolate_patch.dart::249::43)
#1      _SendPort.send (dart::isolate-patch/isolate_patch.dart::230::5)
#2      ServerIsolateChannel.sendRequest (package::analyzer_plugin/src/channel/isolate_channel.dart::249::16)
#3      PluginSession.sendRequest (package::analysis_server/src/plugin/plugin_manager.dart::948::13)
#4      PluginInfo.sendRequest (package::analysis_server/src/plugin/plugin_manager.dart::208::21)
#5      PluginManager.setAnalysisUpdateContentParams (package::analysis_server/src/plugin/plugin_manager.dart::586::14)
#6      LspAnalysisServer._notifyPluginsOverlayChanged (package::analysis_server/src/lsp/lsp_analysis_server.dart::868::21)
#7      LspAnalysisServer._afterOverlayChanged (package::analysis_server/src/lsp/lsp_analysis_server.dart::805::5)
#8      LspAnalysisServer.onOverlayDestroyed (package::analysis_server/src/lsp/lsp_analysis_server.dart::540::5)
#9      TextDocumentCloseHandler.handle.<anonymous closure> (package::analysis_server/src/lsp/handlers/handler_text_document_changes.dart::74::14)
<asynchronous suspension>
#10     LspAnalysisServer._handleNotificationMessage (package::analysis_server/src/lsp/lsp_analysis_server.dart::843::20)
<asynchronous suspension>
#11     LspAnalysisServer.handleMessage.<anonymous closure>.<anonymous closure> (package::analysis_server/src/lsp/lsp_analysis_server.dart::392::15)
<asynchronous suspension>
#12     OperationPerformanceImpl.runAsync (package::analyzer/src/util/performance/operation_performance.dart::164::18)
<asynchronous suspension>
#13     LspAnalysisServer.handleMessage.<anonymous closure> (package::analysis_server/src/lsp/lsp_analysis_server.dart::374::11)
<asynchronous suspension>

:null
```

**Comment by rrousselGit** [comment]
Epic, thanks for investigating!

**Comment by DanTup** [comment]
Not sure what's happening here yet. It seems like we're trying to send a map through to the plugin isolate, and it's failing with this message:

> Invalid argument:: is a regular instance:: Instance of '_HashMap<dynamic, dynamic>'

As far as I can see, the map being sent is valid to send over isolates. The `runtimeType` is `_Map<String, Object>` and the contents of some failing maps are:

```
{
	"id": "2",
	"method": "analysis.updateContent",
	"params": {
		"files": {
			"/Users/danny/Dev/TestStuff/riverpod/packages/riverpod_cli/fixtures/notifiers/golden/change_notifier_provider.dart": {
				"type": "add",
				"content": "import 'package:flutter/material.dart';\n// ignore: unused_import\nimport 'package:flutter_hooks/flutter_hooks.dart';\n// ignore: unnecessary_import\nimport 'package:flutter_riverpod/flutter_riverpod.dart';\nimport 'package:hooks_riverpod/hooks_riverpod.dart';\n// ignore: unnecessary_import\nimport 'package:riverpod/riverpod.dart';\n\nclass Counter extends ChangeNotifier {\n  Counter();\n  int _state = 0;\n  int get state => _state;\n  void increment() {\n    _state++;\n    notifyListeners();\n  }\n\n  void decrement() {\n    _state++;\n    notifyListeners();\n  }\n}\n\nfinal counterProvider = ChangeNotifierProvider((ref) => Counter());\n\nfinal otherProvider = Provider<Counter>((ref) {\n  ref.read(counterProvider);\n  return ref.watch(counterProvider);\n});\n\nclass ConsumerWatch extends ConsumerWidget {\n  const ConsumerWatch({Key? key}) : super(key: key);\n\n  @override\n  Widget build(BuildContext context, ScopedReader watch) {\n    final count = watch(counterProvider);\n    return Center(\n      child: Text('$count'),\n    );\n  }\n}\n\nclass HooksWatch extends HookWidget {\n  const HooksWatch({Key? key}) : super(key: key);\n\n  @override\n  Widget build(BuildContext context) {\n    // ignore: unused_local_variable\n    final countNotifier = useProvider(counterProvider);\n    return Center(\n      child: ElevatedButton(\n        onPressed: () {\n          context.read(counterProvider);\n        },\n        child: const Text('Press Me'),\n      ),\n    );\n  }\n}\n"
			}
		}
	}
}
{
	"id": "3",
	"method": "analysis.updateContent",
	"params": {
		"files": {
			"/Users/danny/Dev/TestStuff/riverpod/packages/riverpod_cli/fixtures/notifiers/golden/change_notifier_provider.dart": {
				"type": "remove"
			}
		}
	}
}
```

Interestingly, this issue has come up before (https://github.com/dart-lang/sdk/issues/24243 / https://github.com/dart-lang/sdk/issues/22982) and at least one comment mentions it being caused by `Map`s inside `Map`s. It seems that the VM has a whitelist for what types are allowed, but it's not clear to me why this is failing.

I tried to reproduce this outside of the server with a small sample (https://gist.github.com/DanTup/a89c3c556b02cf9a042f5e3f4423539e) but everything seems to work fine there. I'll see if I can figure out exactly what the difference between the two is.

**Comment by DanTup** [comment]
@rrousselGit I presume you're _not_ using a stable SDK? I was able to track this down to having `HashMap`s in the payload (and not `LinkedHashMaps`, which are the default). The issue only seems to occur on latest (unreleased) SDKs in my testing though.

I've moved this to https://github.com/dart-lang/sdk/issues/50594 since I think it may be a VM/runtime problem. If you're seeing the same on stable branches, let me know and I'll see whether there might be another issue.

Thanks!
